SixTrack/tools/astuce/astuce++.cpp

```diff
@@ -0,0 +1,708 @@
+	while(CallStorage_itr != CallStorage.end())
+	{
+		DoExpandIF(CallStorage_itr);
+		CallStorage_itr++;
+	}
+
+	//Then +dk
+	std::map<std::string, std::list<LineStorage> >::iterator DeckStorage_itr = DeckStorage.begin();
+	//This loops over each +dk block
+	while(DeckStorage_itr != DeckStorage.end())
+	{
+		DoExpandIF(DeckStorage_itr);
+		DeckStorage_itr++;
+	}
+}
+
+void Astuce::DoExpandIF(std::map<std::string, std::list<LineStorage> >::iterator Storage_itr)
+{
+	std::string BlockName = Storage_itr->first;
+
+	if(OutputFlag)
+	{
+		*LogStream << "Processing: " << BlockName << " with " << Storage_itr->second.size() << " lines." << std::endl;
+	}
+
+	EnabledStates.push(true);
+
+	std::list<LineStorage>::iterator Lines_itr = Storage_itr->second.begin();
+	while(Lines_itr != Storage_itr->second.end())
+	{
+		ProcessIfBlock(Lines_itr);
+		Lines_itr++;
+	}
+
+	EnabledStates.pop();
+
+	if(OutputFlag)
+	{
+		Lines_itr = Storage_itr->second.begin();
+		while(Lines_itr != Storage_itr->second.end())
+		{
+			*LogStream << BlockName << "\t" << Lines_itr->Enabled << " " << Lines_itr->Text << std::endl;
+			Lines_itr++;
+		}
+
+	}
+}
+
+void Astuce::ProcessIfBlock(std::list<LineStorage>::iterator line)
+{
+	//Re-evaluate the enabled state after each +if and +ei statement
+	//Before each +if block we need to store the previous state and restore at +ei (returning to the same depth)
+	//If a parent block is disabled, all children should also be disabled.
+	//Done with a std::stack<bool> and push()/pop()/top()
+
+	//Does this block have a +if statement?
+	if(line->Text.substr(0,3) == "+if")
+	{
+		//Disable this line
+		line->Enabled = false;
+
+		bool EnabledState = false;
+
+		//Find the name of this statement
+		size_t NameStart = line->Text.find_first_not_of(" ", 4);
+		size_t NameEnd = line->Text.find_first_of(" ", NameStart+1);
+		std::string FlagName = FixCase(line->Text.substr(NameStart,NameEnd-NameStart));
+
+		size_t Position = 0;
+		size_t NextOR = 0;
+		size_t NextAND = 0;
+		std::vector<size_t> Splits;
+		std::vector<std::string> FlagNames;
+		std::vector<bool> And;
+		std::vector<bool> Or;
+		std::vector<bool> States;
+
+		Splits.push_back(Position);
+		while(NextOR != std::string::npos && NextAND != std::string::npos)
+		{
+			NextOR = FlagName.find(".OR.", Position);
+			NextAND = FlagName.find(".AND.", Position);
+			Position = std::min(NextOR, NextAND);
+			if(Position != std::string::npos)
+			{
+				Splits.push_back(Position);
+			}
+			Position++;
+		}
+		Splits.push_back(FlagName.size());
+
+		for(size_t n=1; n < Splits.size(); n++)
+		{
+			std::string SplitSubString = FlagName.substr(Splits.at(n-1),Splits.at(n));
+			FlagNames.push_back(SplitSubString);
+		}
+
+		for(size_t n=0; n < FlagNames.size(); n++)
+		{
+			bool NOT = false;
+			bool AND = false;
+			bool OR = false;
+
+			if(FlagNames.at(n).substr(0,5) == ".AND.")
+			{
+				AND = true;
+				FlagNames.at(n).erase(0,5);
+			}
+			if(FlagNames.at(n).substr(0,4) == ".OR.")
```

Comment:
.AND..OR. can't be valid ASTUCE -> this could probably be an `else if`. The "if .not." is however OK , since .OR..NOT. and .AND..NOT. is valid astuce.